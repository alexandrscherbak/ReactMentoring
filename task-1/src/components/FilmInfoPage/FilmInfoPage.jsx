import React from 'react';
import FilmsTable from '../FilmsTable/FilmsTable.jsx';
import Footer from '../Footer/Footer.jsx';
import FilmsByGenrePanel from '../FilmsByGenrePanel/FilmsByGenrePanel.jsx';
import FilmFullInfoContainer from '../FilmFullInfoContainer/FilmFullInfoContainer.jsx';
import MoviesService from '../services/movies.service.js';
import {connect} from 'react-redux';
import SelectedFilmInfoActions from '../../redux/actions/selected-film-info.action.js';
import { withRouter } from "react-router-dom";

export class FilmInfoPage extends React.Component {
	componentWillMount() {
		this.props.selectMovie(this.props.match.params.filmId);
	}

	constructor(props) {
		super(props);
	}

	goToMainPage() {
		this.props.history.push(`/search`);
	}

	selectMovie(filmId) {
		this.props.history.push(`/film/${filmId}`);
		this.props.selectMovie(filmId);
	}

	render() {
		return (
			<React.Fragment>
				{
					this.props.selectedFilmInfo ? 
						<React.Fragment>
							<FilmFullInfoContainer
								searchClicked={() => this.goToMainPage()}
								filmInfo={this.props.selectedFilmInfo}
							/>
							<FilmsByGenrePanel
								genre={this.props.selectedFilmInfo.genres[0]}
							/>
							<FilmsTable
								filmsInfo={this.props.filmsInfo || []}
								selectMovie={movie => this.selectMovie(movie.id)}
							/>
							<Footer/>
						</React.Fragment> :
						null
				}
			</React.Fragment>
		);
	}
}

const mapStateToProps = state => {
	return {
		selectedFilmInfo: state.selectedFilmInfo,
		filmsInfo: state.filmsInfo,
	};
}

const mapDispatchToProps = dispatch => {
	return {
		selectMovie: id => dispatch(SelectedFilmInfoActions.loadSelectedFilmInfo(id)),
	};
}

export default withRouter(connect(
	mapStateToProps,
	mapDispatchToProps,
)(FilmInfoPage));
