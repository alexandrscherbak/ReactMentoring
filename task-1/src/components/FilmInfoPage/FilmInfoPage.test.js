import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { FilmInfoPage } from './FilmInfoPage.jsx';
import renderer from 'react-test-renderer';

Enzyme.configure({ adapter: new Adapter() });

const props = {
	selectedFilmInfo: {
		id: 1,
		poster_path: 'poster_path',
		title: 'film',
		release_date: '2018-04-01',
		genres: ['action', 'drama'],
	},
	filmsInfo: [
		{
			id: 2,
			poster_path: 'poster_path1',
			title: 'film1',
			release_date: '2018-04-01',
			genres: ['action', 'drama'],
		},
		{
			id: 3,
			poster_path: 'poster_path2',
			title: 'film2',
			release_date: '2018-04-01',
			genres: ['action', 'drama'],
		},
	],
	match: {
		params: {
			filmId: 1,
		},
	},
	selectMovie: jest.fn(),
	history: {
		push: jest.fn(),
	}
};

function setup() {
	return mount(<FilmInfoPage {...props}/>);
}

describe('FilmInfoPage', () => {
	test('snapshot', () => {
		const component = renderer.create(
			<FilmInfoPage {...props}/>,
		);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
	it('should call selectMovie in componentDidMount', () => {
		expect(props.selectMovie.mock.calls.length).toBe(1);
	});
	it('should call history.push if FilmFullInfoContainer search clicked', () => {
		const enzymeWrapper = setup();
		const filmFullInfoContainer = enzymeWrapper.find('FilmFullInfoContainer');
		expect(props.history.push.mock.calls.length).toBe(0);
		filmFullInfoContainer.props().searchClicked();
		expect(props.history.push.mock.calls.length).toBe(1);
	});
	it('should call history.push and selectMovie if FilmsTable selectMovie clicked', () => {
		const enzymeWrapper = setup();
		const filmsTable = enzymeWrapper.find('FilmsTable');
		expect(props.history.push.mock.calls.length).toBe(1);
		expect(props.selectMovie.mock.calls.length).toBe(3);
		filmsTable.props().selectMovie({id: 1});
		expect(props.history.push.mock.calls.length).toBe(2);
		expect(props.selectMovie.mock.calls.length).toBe(4);
	});
	it('should not render if no selectedFilmInfo', () => {
		const enzymeWrapper = setup();
		enzymeWrapper.setProps({selectedFilmInfo: null});
		expect(enzymeWrapper.children().length).toBe(0);
	});
});
