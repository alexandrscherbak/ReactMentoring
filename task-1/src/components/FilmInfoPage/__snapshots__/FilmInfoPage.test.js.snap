// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmInfoPage snapshot 1`] = `
Array [
  <div
    className="jsx-1209675933 film-full-info-container"
  >
    <div
      className="jsx-1209675933 film-full-info-header-row"
    >
      <div
        className="jsx-2852891300"
      >
        <span
          className="jsx-2852891300"
        >
          netflixroulette
        </span>
      </div>
      <button
        className="jsx-1209675933"
        onClick={[Function]}
      >
        Search
      </button>
    </div>
    <div
      className="jsx-1209675933 film-full-info"
    >
      <div
        className="jsx-1209675933 film-image"
      >
        <img
          alt="film"
          className="jsx-1209675933"
          src="poster_path"
        />
      </div>
      <div
        className="jsx-1209675933 film-info"
      >
        <h1
          className="jsx-1209675933"
        >
          film
        </h1>
        <div
          className="jsx-1209675933 release-and-rating"
        >
          <span
            className="jsx-1209675933"
          >
            2018
          </span>
          <span
            className="jsx-1209675933"
          />
        </div>
        <span
          className="jsx-1209675933"
        />
      </div>
    </div>
  </div>,
  <div
    className="jsx-4002406560"
  >
    <span
      className="jsx-4002406560"
    >
      Films by 
      action
       genre
    </span>
  </div>,
  <div
    className="jsx-1758079044 films-table"
  >
    <div
      className="jsx-2546762946 film-info-container"
      onClick={[Function]}
    >
      <img
        alt="film1"
        className="jsx-2994430438"
        src="poster_path1"
      />
      <div
        className="jsx-2546762946 title-and-date"
      >
        <div
          className="jsx-433743616"
        >
          <span
            className="jsx-433743616"
          >
            film1
          </span>
        </div>
        <div
          className="jsx-617062702"
        >
          <span
            className="jsx-617062702"
          >
            2018
          </span>
        </div>
      </div>
      <div
        className="jsx-672986837"
      >
        <span
          className="jsx-672986837"
        >
          action
        </span>
      </div>
    </div>
    <div
      className="jsx-2546762946 film-info-container"
      onClick={[Function]}
    >
      <img
        alt="film2"
        className="jsx-2994430438"
        src="poster_path2"
      />
      <div
        className="jsx-2546762946 title-and-date"
      >
        <div
          className="jsx-433743616"
        >
          <span
            className="jsx-433743616"
          >
            film2
          </span>
        </div>
        <div
          className="jsx-617062702"
        >
          <span
            className="jsx-617062702"
          >
            2018
          </span>
        </div>
      </div>
      <div
        className="jsx-672986837"
      >
        <span
          className="jsx-672986837"
        >
          action
        </span>
      </div>
    </div>
  </div>,
  <div
    className="jsx-429413428"
  >
    <div
      className="jsx-2852891300"
    >
      <span
        className="jsx-2852891300"
      >
        netflixroulette
      </span>
    </div>
  </div>,
]
`;
